File: app/nucleo/motor.js
/* Motor de projeção: transforma aportes + taxa em uma trajetória mês a mês.

   Convenções (documentadas em docs/03-motor-de-calculo.md):
     - O mês 1 é o primeiro mês cheio de rendimento.
     - O capital inicial entra no começo do mês 1 e rende os N meses.
     - Cada aporte mensal entra no começo do seu mês e rende daquele mês em diante.
     - Cada entrada de dinheiro vira um LOTE próprio, porque a tabela regressiva
       de IR conta o prazo de cada aporte separadamente.
     - Prazo para fins de IR usa mês comercial de 30 dias. */

import { resolver, arredondar } from './indexadores.js';
import { tributar, DIAS_MES_COMERCIAL } from './tributos.js';

export const MAX_MESES = 600;

export function projetar(ativo, indicadores, premissas, {
  valor_inicial: valorInicial = 0,
  aporte_mensal: aporteMensal = 0,
  meses = 12,
  cenario = 'base',
  considerar_ir: considerarIr = true,
  considerar_inflacao: considerarInflacao = true,
  considerar_valorizacao_projetada: considerarValorizacaoProjetada = false,
} = {}, fundos = {}) {
  if (!Number.isFinite(meses) || meses < 1 || meses > MAX_MESES) {
    throw new Error(`Prazo deve estar entre 1 e ${MAX_MESES} meses`);
  }
  if (valorInicial < 0 || aporteMensal < 0) {
    throw new Error('Valores de aporte não podem ser negativos');
  }
  if (valorInicial <= 0 && aporteMensal <= 0) {
    throw new Error('Informe um valor inicial ou um aporte mensal');
  }

  const taxa = resolver(ativo, indicadores, premissas, cenario, fundos, considerarValorizacaoProjetada);
  const taxaAm = taxa.taxa_am;
  // Só FII (regime 'fii') preenche estes dois campos, separando o dividendo
  // (isento, não compõe o capital que sofre ganho de capital) da valorização
  // patrimonial (tributável na venda). Para todo outro ativo, isentoAm é 0 e
  // tributavelAm é a taxa cheia - o cálculo abaixo se reduz exatamente ao que
  // era antes desta separação existir (ver docs/03-motor-de-calculo.md).
  const isentoAm = taxa.componente_isento_am || 0;
  const tributavelAm = taxa.componente_tributavel_am ?? taxaAm;

  const lotes = [];
  const serie = [{ mes: 0, bruto: arredondar(valorInicial), investido: arredondar(valorInicial) }];
  let investido = valorInicial;
  let dividendosIsentos = 0;

  if (valorInicial > 0) {
    lotes.push({ mes_entrada: 1, principal: valorInicial, valor_final: valorInicial });
  }

  for (let mes = 1; mes <= meses; mes++) {
    if (aporteMensal > 0) {
      lotes.push({ mes_entrada: mes, principal: aporteMensal, valor_final: aporteMensal });
      investido += aporteMensal;
    }
    for (const lote of lotes) {
      dividendosIsentos += lote.valor_final * isentoAm;
      lote.valor_final *= 1 + tributavelAm;
    }
    const brutoPatrimonialMes = lotes.reduce((s, l) => s + l.valor_final, 0);
    serie.push({
      mes,
      bruto: arredondar(brutoPatrimonialMes + dividendosIsentos),
      investido: arredondar(investido),
    });
  }

  // `brutoPatrimonial` é só a parte que ainda está "na cota" - é sobre ela
  // que o IR de ganho de capital incide. Os dividendos já foram recebidos
  // (isentos) mês a mês e não entram nos lotes.
  const brutoPatrimonial = lotes.reduce((s, l) => s + l.valor_final, 0);
  const brutoFinal = brutoPatrimonial + dividendosIsentos;
  for (const lote of lotes) {
    lote.dias = (meses - lote.mes_entrada + 1) * DIAS_MES_COMERCIAL;
  }

  const impostos = considerarIr
    ? tributar(ativo.tributacao.regime, lotes, brutoPatrimonial, premissas.tributacao)
    : { iof: 0, ir: 0, total: 0, aliquota_efetiva: 0, detalhe: 'Impostos desligados nesta simulação' };

  const liquido = brutoPatrimonial - impostos.total + dividendosIsentos;
  const inflacao = calcularInflacao(premissas, indicadores, cenario, meses);
  const liquidoReal = considerarInflacao ? liquido / inflacao.fator : liquido;

  return {
    ativo_id: ativo.id,
    nome: ativo.nome,
    classe: ativo.classe,
    risco: ativo.risco,
    taxa,
    meses,
    investido: arredondar(investido),
    bruto: arredondar(brutoFinal),
    rendimento_bruto: arredondar(brutoFinal - investido),
    dividendos_isentos: arredondar(dividendosIsentos),
    impostos,
    liquido: arredondar(liquido),
    rendimento_liquido: arredondar(liquido - investido),
    liquido_real: arredondar(liquidoReal),
    ganho_real: arredondar(liquidoReal - investido),
    rentabilidade_bruta_pct: variacao(brutoFinal, investido),
    rentabilidade_liquida_pct: variacao(liquido, investido),
    rentabilidade_real_pct: variacao(liquidoReal, investido),
    inflacao,
    serie,
    lotes: lotes.length,
  };
}

export function comparar(ativos, indicadores, premissas, parametros, fundos = {}) {
  const resultados = [];
  const erros = [];

  for (const ativo of ativos) {
    try {
      resultados.push(projetar(ativo, indicadores, premissas, parametros, fundos));
    } catch (erro) {
      erros.push({ ativo_id: ativo.id, erro: erro.message });
    }
  }

  resultados.sort((a, b) => b.liquido - a.liquido);
  resultados.forEach((r, i) => {
    r.ranking = i + 1;
    r.diferenca_para_1o = arredondar(r.liquido - resultados[0].liquido);
  });

  return { resultados, erros };
}

function calcularInflacao(premissas, indicadores, cenario, meses) {
  const bloco = (premissas.estimativas || {}).ipca_projetado_aa;
  let ipcaAa;
  if (bloco && typeof bloco === 'object') {
    ipcaAa = Number(bloco[cenario] ?? bloco.base ?? indicadores.ipca_12m);
  } else {
    ipcaAa = Number(bloco || indicadores.ipca_12m);
  }
  const fator = Math.pow(1 + ipcaAa / 100, meses / 12);
  return {
    ipca_aa_projetado: arredondar(ipcaAa, 4),
    fator,
    perda_poder_compra_pct: arredondar((1 - 1 / fator) * 100),
    cenario,
  };
}

function variacao(final, investido) {
  if (investido <= 0) return 0;
  return arredondar((final / investido - 1) * 100);
}

File: backend/engine/motor.py
"""Motor de projecao: transforma aportes + taxa em uma trajetoria mes a mes.

Convencoes (documentadas em docs/03-motor-de-calculo.md):
  - O mes 1 e o primeiro mes cheio de rendimento.
  - O capital inicial entra no comeco do mes 1 e rende os N meses.
  - Cada aporte mensal entra no comeco do seu mes e rende daquele mes em diante.
  - Cada entrada de dinheiro vira um LOTE proprio, porque a tabela regressiva
    de IR conta o prazo de cada aporte separadamente.
  - Prazo para fins de IR usa mes comercial de 30 dias.
"""
from __future__ import annotations

from . import indexadores, tributos

MAX_MESES = 600


def projetar(
    ativo: dict,
    indicadores_atuais: dict,
    premissas: dict,
    valor_inicial: float,
    aporte_mensal: float,
    meses: int,
    cenario: str = "base",
    considerar_ir: bool = True,
    considerar_inflacao: bool = True,
) -> dict:
    if meses < 1 or meses > MAX_MESES:
        raise ValueError(f"Prazo deve estar entre 1 e {MAX_MESES} meses")
    if valor_inicial < 0 or aporte_mensal < 0:
        raise ValueError("Valores de aporte nao podem ser negativos")
    if valor_inicial <= 0 and aporte_mensal <= 0:
        raise ValueError("Informe um valor inicial ou um aporte mensal")

    taxa = indexadores.resolver(ativo, indicadores_atuais, premissas, cenario)
    taxa_am = taxa["taxa_am"]

    lotes = []
    serie = [{"mes": 0, "bruto": round(valor_inicial, 2), "investido": round(valor_inicial, 2)}]
    investido = valor_inicial
    if valor_inicial > 0:
        lotes.append({"mes_entrada": 1, "principal": valor_inicial, "valor_final": valor_inicial})

    for mes in range(1, meses + 1):
        if aporte_mensal > 0:
            lotes.append({"mes_entrada": mes, "principal": aporte_mensal, "valor_final": aporte_mensal})
            investido += aporte_mensal

        for lote in lotes:
            lote["valor_final"] *= 1 + taxa_am

        bruto = sum(l["valor_final"] for l in lotes)
        serie.append({"mes": mes, "bruto": round(bruto, 2), "investido": round(investido, 2)})

    bruto_final = sum(l["valor_final"] for l in lotes)
    rendimento_bruto = bruto_final - investido

    for lote in lotes:
        lote["dias"] = (meses - lote["mes_entrada"] + 1) * tributos.DIAS_MES_COMERCIAL

    regime = ativo["tributacao"]["regime"]
    if considerar_ir:
        imposto = tributos.tributar(regime, lotes, bruto_final, premissas["tributacao"])
    else:
        imposto = {"iof": 0.0, "ir": 0.0, "total": 0.0, "aliquota_efetiva": 0.0,
                   "detalhe": "Impostos desligados nesta simulacao"}

    liquido = bruto_final - imposto["total"]

    inflacao = _inflacao(premissas, indicadores_atuais, cenario, meses)
    liquido_real = liquido / inflacao["fator"] if considerar_inflacao else liquido

    return {
        "ativo_id": ativo["id"],
        "nome": ativo["nome"],
        "classe": ativo["classe"],
        "risco": ativo.get("risco"),
        "taxa": taxa,
        "meses": meses,
        "investido": round(investido, 2),
        "bruto": round(bruto_final, 2),
        "rendimento_bruto": round(rendimento_bruto, 2),
        "impostos": imposto,
        "liquido": round(liquido, 2),
        "rendimento_liquido": round(liquido - investido, 2),
        "liquido_real": round(liquido_real, 2),
        "ganho_real": round(liquido_real - investido, 2),
        "rentabilidade_bruta_pct": _pct(bruto_final, investido),
        "rentabilidade_liquida_pct": _pct(liquido, investido),
        "rentabilidade_real_pct": _pct(liquido_real, investido),
        "inflacao": inflacao,
        "serie": serie,
        "lotes": len(lotes),
    }


def comparar(
    ativos: list,
    indicadores_atuais: dict,
    premissas: dict,
    **parametros,
) -> dict:
    resultados = []
    erros = []
    for ativo in ativos:
        try:
            resultados.append(projetar(ativo, indicadores_atuais, premissas, **parametros))
        except (ValueError, KeyError) as erro:
            erros.append({"ativo_id": ativo.get("id"), "erro": str(erro)})

    resultados.sort(key=lambda r: r["liquido"], reverse=True)
    for posicao, resultado in enumerate(resultados, start=1):
        resultado["ranking"] = posicao
        if resultados:
            resultado["diferenca_para_1o"] = round(resultado["liquido"] - resultados[0]["liquido"], 2)

    return {"resultados": resultados, "erros": erros}


def _inflacao(premissas: dict, indicadores_atuais: dict, cenario: str, meses: int) -> dict:
    bloco = premissas.get("estimativas", {}).get("ipca_projetado_aa", {})
    if isinstance(bloco, dict):
        ipca_aa = float(bloco.get(cenario, bloco.get("base", indicadores_atuais["ipca_12m"])))
    else:
        ipca_aa = float(bloco or indicadores_atuais["ipca_12m"])
    fator = (1 + ipca_aa / 100) ** (meses / 12)
    return {
        "ipca_aa_projetado": round(ipca_aa, 4),
        "fator": fator,
        "perda_poder_compra_pct": round((1 - 1 / fator) * 100, 2),
        "cenario": cenario,
    }


def _pct(final: float, investido: float) -> float:
    if investido <= 0:
        return 0.0
    return round((final / investido - 1) * 100, 2)
